### PR TITLE
Dev 64080:  Delegated authentication feature - Screens to Approve & Register

### PR DESCRIPTION
**Pull request by robertdalmeida**
## Summary
Support screens to register and approve in the delegated authentication flow.

- [ ] Ready to be merged in the next release

Additionally updated to the feature branch: 
- Consuming LinkTextView

<new> Delegated authentication intermediate screens  </new>

### Register for DA
https://github.com/Adyen/adyen-ios/assets/96450350/90c01739-e27e-4882-959f-8cb98b7e993d

### Approval using DA
https://github.com/Adyen/adyen-ios/assets/96450350/9885868e-d651-4f9a-a2dd-3db61674f26a

### Delete credentials
https://github.com/Adyen/adyen-ios/assets/96450350/9bcf2c82-031c-48b5-a972-08d0b297c41e

### Timeout
https://github.com/Adyen/adyen-ios/assets/96450350/4fc6061d-c2ee-4fc0-99d7-2e7236bd231b

### CODE REVIEW

File: AdyenActions/Components/3DS2/Action handlers/3DS2+Delegated Authentication/ThreeDS2PlusDACoreActionHandler.swift [hunk 1]
```diff
@@ -148,6 +161,108 @@
...
+            
+            guard let delegatedAuthenticationInput = fingerprintToken.delegatedAuthenticationSDKInput else {
+                failureHandler(ThreeDS2PlusDACoreActionError.sdkInputNotAvailableForApproval)
+                return
+            }
+            
+            isDeviceRegisteredForDelegatedAuthentication(
+                delegatedAuthenticationInput: delegatedAuthenticationInput,
+                registeredHandler: { [weak self] in
+                    guard let self = self else { return }
```
Comment (goergisn) [inline AdyenActions/Components/3DS2/Action handlers/3DS2+Delegated Authentication/ThreeDS2PlusDACoreActionHandler.swift]:
```suggestion
                    guard let self else { return }
```

File: AdyenActions/Components/3DS2/Action handlers/3DS2+Delegated Authentication/ThreeDS2PlusDACoreActionHandler.swift [hunk 2]
```diff
@@ -148,6 +161,108 @@
...
+        }
+        
+        // MARK: Delegated Authentication Approval
+        
+        private func showApprovalScreenWhenDeviceIsRegistered(delegatedAuthenticationInput: String,
+                                                              completion: @escaping (Result<String, DelegateAuthenticationError>) -> Void,
+                                                              failureHandler: @escaping (Error) -> Void) {
+            presenter.showApprovalScreen(component: self,
+                                         approveAuthenticationHandler: { [weak self] in
+                                             guard let self = self else { return }
```
Comment (goergisn) [inline AdyenActions/Components/3DS2/Action handlers/3DS2+Delegated Authentication/ThreeDS2PlusDACoreActionHandler.swift]:
```suggestion
                                             guard let self else { return }
```

File: AdyenActions/Components/3DS2/Action handlers/3DS2+Delegated Authentication/ThreeDS2PlusDACoreActionHandler.swift [hunk 3]
```diff
@@ -148,6 +161,108 @@
...
+        }
+        
+        private func isDeviceRegisteredForDelegatedAuthentication(delegatedAuthenticationInput: String,
+                                                                  registeredHandler: @escaping () -> Void,
+                                                                  notRegisteredHandler: @escaping (Error) -> Void) {
+            delegatedAuthenticationService.isDeviceRegistered(withAuthenticationInput: delegatedAuthenticationInput) { result in
+                switch result {
+                case let .failure(error):
+                    notRegisteredHandler(error)
+                case let .success(success):
```
Comment (goergisn) [inline AdyenActions/Components/3DS2/Action handlers/3DS2+Delegated Authentication/ThreeDS2PlusDACoreActionHandler.swift]:
```suggestion
                case let .success(isRegistered):
```

File: AdyenActions/Components/3DS2/Action handlers/3DS2+Delegated Authentication/ThreeDS2PlusDACoreActionHandler.swift [hunk 4]
```diff
@@ -148,6 +161,108 @@
...
+                    if success {
```
Comment (goergisn) [inline AdyenActions/Components/3DS2/Action handlers/3DS2+Delegated Authentication/ThreeDS2PlusDACoreActionHandler.swift]:
```suggestion
                    if isRegistered {
```

File: AdyenActions/Components/3DS2/Action handlers/3DS2+Delegated Authentication/ThreeDS2PlusDACoreActionHandler.swift [hunk 5]
```diff
@@ -148,6 +161,108 @@
...
+                    }
+                }
+            }
+        }
+                
+        // MARK: Delegated Authentication Registration
+
+        internal var shouldShowRegistrationScreen: Bool {
+            delegatedAuthenticationState.isDeviceRegistrationFlow
+                && presenter.userInput != .approveDifferently
```
Comment (goergisn) [inline AdyenActions/Components/3DS2/Action handlers/3DS2+Delegated Authentication/ThreeDS2PlusDACoreActionHandler.swift]: Maybe better use a switch here (or in a computed property on `ThreeDS2PlusDAScreenUserInput`) so we get a compile time error when a new userInput is added that might need changes here.

File: AdyenActions/Components/3DS2/Action handlers/3DS2+Delegated Authentication/ThreeDS2PlusDACoreActionHandler.swift [hunk 6]
```diff
@@ -175,44 +290,49 @@
...
+                                              delegatedAuthenticationSDKOutput: result.successResult,
+                                              deleteDelegatedAuthenticationCredentials: nil,
+                                              completionHandler: completionHandler)
+                            }
+                        },
+                        fallbackHandler: {
+                            completionHandler(.success(challengeResult))
+                        }
+                    )
+                } else if presenter.userInput == .deleteDA {
```
Comment (goergisn) [inline AdyenActions/Components/3DS2/Action handlers/3DS2+Delegated Authentication/ThreeDS2PlusDACoreActionHandler.swift]: Maybe also good to not compare against a specific case but have a computed property `presenter.userInput.shouldDeleteDelegatedAuthenticationCredentials` (or something shorter) with a switch to not have any new added user input to slip through accidentally

File: AdyenActions/Components/3DS2/Action handlers/3DS2+Delegated Authentication/ThreeDS2PlusDACoreActionHandler.swift [hunk 7]
```diff
@@ -175,44 +290,49 @@
...
+                    completionHandler(.success(challengeResult))
+                    return
+                }
+                if shouldShowRegistrationScreen {
+                    presenter.showRegistrationScreen(
+                        component: self,
+                        registerDelegatedAuthenticationHandler: { [weak self] in
+                            guard let self = self else { return }
+                            self.performDelegatedRegistration(sdkInput) { [weak self] result in
+                                self?.deliver(challengeResult: challengeResult,
```
Comment (goergisn) [inline AdyenActions/Components/3DS2/Action handlers/3DS2+Delegated Authentication/ThreeDS2PlusDACoreActionHandler.swift]: Might be nice to reduce the nesting a bit here as it gets quite deep

File: AdyenActions/UI/View Controllers/DelegatedAuthentication/DelegatedAuthenticationView.swift [hunk 8]
```diff
@@ -0,0 +1,210 @@
...
+//
+// This file is open source and available under the MIT license. See the LICENSE file for more info.
+//
+
+@_spi(AdyenInternal) import Adyen
+import Foundation
+import UIKit
+
+internal protocol DelegatedAuthenticationViewDelegate: AnyObject {
+    func firstButtonTapped()
```
Comment (neelSharma12) [inline AdyenActions/UI/View Controllers/DelegatedAuthentication/DelegatedAuthenticationView.swift:12]: Which buttons are these?

File: AdyenActions/Components/3DS2/Action handlers/3DS2+Delegated Authentication/ThreeDS2PlusDACoreActionHandler.swift (hunk 5 above)
Comment (robertdalmeida) [reply AdyenActions/Components/3DS2/Action handlers/3DS2+Delegated Authentication/ThreeDS2PlusDACoreActionHandler.swift]: Good idea, created a computed property at ThreeDS2PlusDAScreenUserInput. Thanks.

File: AdyenActions/Components/3DS2/Action handlers/3DS2+Delegated Authentication/ThreeDS2PlusDACoreActionHandler.swift (hunk 6 above)
Comment (robertdalmeida) [reply AdyenActions/Components/3DS2/Action handlers/3DS2+Delegated Authentication/ThreeDS2PlusDACoreActionHandler.swift]: Sure. Created a computed property for this. Thanks.

File: AdyenActions/Components/3DS2/Action handlers/3DS2+Delegated Authentication/ThreeDS2PlusDACoreActionHandler.swift (hunk 7 above)
Comment (robertdalmeida) [reply AdyenActions/Components/3DS2/Action handlers/3DS2+Delegated Authentication/ThreeDS2PlusDACoreActionHandler.swift]: Ahh.. the closures. Moved a few levels into into another method.

File: AdyenActions/UI/View Controllers/DelegatedAuthentication/DelegatedAuthenticationView.swift (hunk 8 above)
Comment (robertdalmeida) [reply AdyenActions/UI/View Controllers/DelegatedAuthentication/DelegatedAuthenticationView.swift:12]: `DelegatedAuthenticationView` is shared across both the registration and the approval screens. 
The first button naming is to be agnostic to how they are used in each of the screens. "Use biometrics" & "Approve" respectively.

**Comment by robertdalmeida** [comment]
Creating another MR as a continuation to this one due to branch naming issues - https://github.com/Adyen/adyen-ios/pull/1549

**Pull request closed without merging**